new service: version (WAS: introspection)

AkihiroSuda opened the pull request:
Initial introspection service (#745)

@stevvooe @Random-Liu 

e.g.

    $ ctr introspect
    (*introspection.IntrospectResponse)({
            Version: (*introspection.Version)({
                    Version: (string) (len=15) "1.0-dev+unknown",
                    GitCommit: (string) (len=9) "12d65ce.m"
            })
    })

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

mlaventure commented:
I would have expected a `Version` rpc call to get the version.

The `Introspect` name would be better used to gather information about a given runtime/service I think.

Otherwise, have the `IntrospectRequest` hold a `type` field and the response be a serie of `oneof` fields (one for each type).

crosbymichael commented:
@AkihiroSuda thanks for working on this, we'll review today

estesp commented:
My immediate thought was the same as @mlaventure ; version seems like it should be just a `Version` call added to rpc definition.

Random-Liu commented:
@AkihiroSuda Thanks for working on this!

Also agree with @mlaventure, docker also has separate version and introspection api (`docker version` and `docker info`), which is reasonable to me. :)

stevvooe commented:
I'm not sure about adding a reported version API. I really don't want to see clients having to do things like this:

```
if version > 2 {
  // do this
} else {
  // do that
}
```

Such a thing encumbers non-standard implementations. Instead, functionality detection should be done through feature tests and service queries.

In general, I am fine with reporting this info via the API, but I don't want it to become a method for detecting whether or not containerd will do something. That approach creates an endless mess.

Random-Liu commented:
> In general, I am fine with reporting this info via the API, but I don't want it to become a method for detecting whether or not containerd will do something. That approach creates an endless mess.

Have to admit that, we are using docker version to figure out whether a feature is supported or not in Kubernetes... Because there is not better solution for us given that People are using the same Kubernetes version with different docker version. :p 

But except that, we also need to include the runtime name and runtime version in the node information, so that user knows what runtime is running on the node.

stevvooe commented:
> @stevvooe Have to admit that, we are indeed using docker version to figure out whether a feature is supported or not in Kubernetes... Because there is no better solution for us given that People are using the same Kubernetes version with different docker versions. :p

Yes, I have seen this pain first-hand. Hopefully we can avoid it. More concretely, we can avoid it by versioning the services themselves (ie `containerd.v1.ContentService`). 

> But except that, the more important requirement is that we need to include the runtime name and runtime version in the node information, so that user knows what runtime is running on that node.

Agreed. This is the primary use case of this feature.

samuelkarp commented:
> Have to admit that, we are indeed using docker version to figure out whether a feature is supported or not in Kubernetes... Because there is no better solution for us given that People are using the same Kubernetes version with different docker versions. :p

We do the same thing for Amazon ECS, though we use the Docker Remote API version instead of the daemon version.

> Yes, I have seen this pain first-hand. Hopefully we can avoid it. More concretely, we can avoid it by versioning the services themselves (ie containerd.v1.ContentService).

That's one approach.  Another approach could involve containerd reporting a set of capabilities that an orchestrator can use to determine whether a given feature is supported or not.  I like the capability model a bit better, because it better represents what an orchestrator needs without overloading the version concept.

AkihiroSuda commented:
Updated 

- added `Plugins` so as to detect available plugins
- added `Version.{Major, Minor, Patch}` integers for comparability
- no `Version{Request,Response}` yet, we can add them later on demand

example:
```
$ ctr introspect
(*introspection.IntrospectResponse)({
        Version: (*introspection.Version)({
                Major: (int32) 1,
                Minor: (int32) 0,
                Patch: (int32) 0,
                Suffix: (string) (len=4) "-dev",
                GitCommit: (string) (len=40) "4b071cd509194f5c2c334997b13b65d554d0f790",
                GitDirty: (bool) false
        }),
        Plugins: ([]*introspection.PluginRegistration) (len=11 cap=16) {
                (*introspection.PluginRegistration)({
                        Name: (string) (len=11) "rootfs-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=7) "cgroups",
                        Type: (introspection.PluginType) 4
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=16) "snapshot-overlay",
                        Type: (introspection.PluginType) 3
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=12) "content-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=12) "runtime-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=11) "images-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=5) "linux",
                        Type: (introspection.PluginType) 1
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=14) "snapshot-btrfs",
                        Type: (introspection.PluginType) 3
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=16) "healthcheck-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=18) "introspection-grpc",
                        Type: (introspection.PluginType) 2
                }),
                (*introspection.PluginRegistration)({
                        Name: (string) (len=12) "metrics-grpc",
                        Type: (introspection.PluginType) 2
                })
        }
})
```

mlaventure commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,39 @@
+	int32 minor = 2;
+	int32 patch = 3;
+	string suffix = 4;
+	string git_commit = 5;
```
`s/git_commit/revision/`, not point in making it tool specific.

mlaventure commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,39 @@
+	int32 patch = 3;
+	string suffix = 4;
+	string git_commit = 5;
+	bool git_dirty = 6;
```
`s/git_dirty/dirty/`

stevvooe replied:
Let's just include this in the commit spec.

mlaventure commented:
> have the IntrospectRequest hold a type field and the response be a serie of oneof fields (one for each type).

This would make it extensible without having to pack everything in one message I think.

```
message IntrospectRequest {
   TypeEnum type = 1;
   string name = 2;
}
```

the availability of plugins and version could be returned when requestion `"containerd"` itself.

But I could also use it to query capabilities of a runtime plugin for instance (e.g. support for pause,resume,checkpoint etc) which would fit with @samuelkarp suggestion.

WDYT?

Random-Liu commented:
> We do the same thing for Amazon ECS, though we use the Docker Remote API version instead of the daemon version.

Actually, we are also using remote api version. We get the api version from `docker version` (actually `dockerclient.Version()`)

stevvooe commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,39 @@
+message Version {
+	int32 major = 1;
+	int32 minor = 2;
+	int32 patch = 3;
```
Please don't break down the version field like this. Just have a string.

AkihiroSuda commented:
updated PR
- `GitCommit` -> `VCSRevision`
- include `GitDirty` to `VCSRvision` (as `${revision}-unsupported`)
- add `Version` string to the grpc response. still numeric major.minor.patch are reserved for ease of comparison
- add selector (`ctr introspect --type all | version | plugins`). currently, only `--type` is implemented. `--name` should be another PR.

```console
$ ./bin/ctr introspect --type version
(*introspection.IntrospectResponse)({
        Version: (*introspection.Version)({
                Version: (string) (len=9) "1.0.0-dev",
                Major: (int32) 1,
                Minor: (int32) 0,
                Patch: (int32) 0,
                Suffix: (string) (len=4) "-dev",
                VCSRevision: (string) (len=52) "71b31840dded08b3b8a354491a203093e58f5195-unsupported"
        }),
        Plugins: ([]*introspection.PluginRegistration) <nil>
})
```

stevvooe commented:
> still numeric major.minor.patch are reserved for ease of comparison

Really, please remove these. All of these information just confuses things. All you need is a version string. As I said, we don't want people comparing version strings for feature detection. It is a mess of a disaster every time.

AkihiroSuda commented:
@stevvooe sure, removed them

stevvooe commented on Makefile:
```diff
@@ -4,8 +4,10 @@ ROOTDIR=$(dir $(abspath $(lastword $(MAKEFILE_LIST))))
 DESTDIR=/usr/local
 
-# Used to populate version variable in main package.
-VERSION=$(shell git describe --match 'v[0-9]*' --dirty='.m' --always)
```
Please don't remove this. We are going to use tags to set the version.

Do you want us to setup a 1.0.0-dev tag? I was going to do this when submitting the versioning plan.

stevvooe commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,47 @@
+
+message Version {
+	string version = 1;
+	string vcs_revision = 2 [(gogoproto.customname) = "VCSRevision"];
```
Let's just call this "revision".

stevvooe commented on version.go:
```diff
@@ -1,12 +1,13 @@
 	Package = "github.com/containerd/containerd"
 
-	// Version holds the complete version number.
-	Version = "1.0-dev+unknown"
```
Leave the `unknown` here. It signifies that the linker didn't actually fill in the version string and the that the version is unknown.

stevvooe commented on cmd/ctr/introspect.go:
```diff
@@ -0,0 +1,48 @@
+		sel := &introspection.IntrospectionSelector{}
+		switch typ := strings.ToUpper(context.String("type")); typ {
+		case "ALL", "":
+			sel.Type = introspection.IntrospectionSelectorType_ALL
```
There is a plugin in gogoproto to make these generate correctly. `enum_customname` and `enumvalue_customname` will allow these to follow Go style guidelines for constants. swarmkit has this done correctly in a few places.

stevvooe commented on version.go:
```diff
@@ -1,12 +1,13 @@
+	// VCSRevision is filled with the VCS (e.g. git) revision being used to build the
 	// program at linking time
-	GitCommit = ""
+	VCSRevision = ""
```
`Revision`.

stevvooe commented on version.go:
```diff
@@ -1,12 +1,13 @@
 package containerd
```
@crosbymichael Any objections about moving this to a separate package?

crosbymichael replied:
Nope

AkihiroSuda commented:
updated PR

stevvooe commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,47 @@
+	PluginType type = 2;
+}
+
+enum IntrospectionSelectorType {
```
Let's just make separate methods for this, rather than implementing an adhoc mechanism.

AkihiroSuda commented:
updated PR
removed plugins from the introspection result, in favor of Samuel's capability proposal

crosbymichael commented:
I think right now we may want to take it slow and just add a `version` service that reports back the version.  Introspection is one of those overloaded terms that will cause a service to be bloated very quickly.

There is also container level introspection, daemon level, etc...  Lets make sure that the services remain small and we know what the scope of each service is.

crosbymichael commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,25 @@
+
+import "gogoproto/gogo.proto";
+
+service Introspection {
```
Change to Version

crosbymichael commented on cmd/ctr/introspect.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+var introspectCommand = cli.Command{
+	Name:      "introspect",
```
Make versionCommand

crosbymichael commented on cmd/ctr/utils.go:
```diff
@@ -129,3 +139,11 @@ func stopCatch(sigc chan os.Signal) {
 	close(sigc)
 }
+
+func getSpewDumper() *spew.ConfigState {
```
Do we need a new spew dep when we are just going to print out two strings?

crosbymichael commented on api/services/introspection/introspection.proto:
```diff
@@ -0,0 +1,25 @@
+import "gogoproto/gogo.proto";
+
+service Introspection {
+	rpc IntrospectVersion(IntrospectVersionRequest) returns (Version);
```
We probably only need one RPC, `Version` or something similar

AkihiroSuda commented:
updated PR

```console
$ ctr version
Client:
  Version: a94fc8f
  Revision: a94fc8f3690aa50c1811951feca96abaa065b66e

Server:
  Version: a94fc8f
  Revision: a94fc8f3690aa50c1811951feca96abaa065b66e
```

stevvooe commented on api/services/version/version.proto:
```diff
@@ -0,0 +1,15 @@
+import "gogoproto/gogo.proto";
+
+service VersionService {
+	rpc GetVersion(google.protobuf.Empty) returns (Version);
```
No need for `Get` here.

stevvooe commented on api/services/version/version.proto:
```diff
@@ -0,0 +1,15 @@
+import "google/protobuf/empty.proto";
+import "gogoproto/gogo.proto";
+
+service VersionService {
```
Just call it `Version`. The go code will name it a service for you.

estesp commented:
Needs rebase; agree with @stevvooe on the two minor changes to names--after that should be good to merge, I think

AkihiroSuda commented:
@stevvooe 

`make generate` unhappy
```console
$ cat api/services/version/version.proto
syntax = "proto3";

package containerd.v1;

import "google/protobuf/empty.proto";
import "gogoproto/gogo.proto";

service Version {
        rpc Version(google.protobuf.Empty) returns (Version);
}

message Version {
        string version = 1;
        string revision = 2;
}
$ make generate
2017/05/10 05:03:14 github.com/containerd/containerd/api/services/version/version.proto:8:9: "Version" is already defined in "containerd.v1".
github.com/containerd/containerd/api/services/version/version.proto:9:53: "Version" is not a message type.
 exit status 1
```

stevvooe commented:
@AkihiroSuda Call `message Version`, `VersionResponse`.

AkihiroSuda commented:
done

stevvooe commented:
LGTM
